odoo/addons/base/controllers/avatar.py

```diff
@@ -0,0 +1,82 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from hashlib import sha1
+
+from odoo.http import Controller, request, route, Response, local_redirect
+from odoo import tools
+
+class Avatar(Controller):
+    @route([
+        '/web/avatar/<string:res_model>/<int:user_id>',
+        '/web/avatar/<string:res_model>/<int:user_id>/<string:field>'
+    ], auth='public', methods=['GET'])
+    def get_avatar(self, res_model='res.users', user_id=1, field='image_128', **kwargs):
+        """
+        Returns a custom avatar based on the user's data
+            res_model: corresponding model
+            user_id: user id
+            field: image field
+        """
+        if res_model not in ['res.users', 'res.partner', 'hr.employee', 'hr.employee.public']:
+            return local_redirect(f'/web/image?model={res_model}&id={user_id}&field={field}')
+        user = request.env[res_model].sudo().browse(user_id).exists()
+        if not user:
+            return request.not_found()
+        if user[field]:
+            return local_redirect(f'/web/image?model={res_model}&id={user_id}&field={field}')
+        avatar = _generate_avatar_svg(user.name)
+        return Response(avatar, mimetype='image/svg+xml')
+
+def _encode_name(name):
+    return sha1(name.encode('utf-8'))
```

Comment:
Just for the sake of the security mailbox, could you use sha2 instead of sha1 ? There are people in the community who send security reports whenever they read we use an older hashing algorithm, even when there is actually no security issue. sha512 is the fastest variant. It's too bad python lacks a deterministic general purpose hashing function :(

Reply:
For the record, [`binascii.crc32`](https://docs.python.org/3/library/binascii.html#binascii.crc32) and [`zip.adler32`](https://docs.python.org/3/library/zlib.html#zlib.adler32) are two checksum functions, some people argue online we could use them to hash stuff. They are wrong, they have poor performances when it comes to avoid collisions. `hashlib` *is* the best way we have to hash stuff the deterministic way.